### ISSUE TITLE
Performance of beta 1.9 beta 3 much slower than previous public release

### ISSUE THREAD

**Issue by brendand** [abbeycode/UnzipKit#75]
### Steps to reproduce issue (Detailed enough for me to reproduce):

When using the pre-built binary on macOS High Sierra, the current 1.9.3 beta version is significantly slower at compressing and de-compressing than the previous version I was using (from many months ago).

### What you expect to happen:

I expected the performance to be on-par at least.

### What actually happened:

It was significantly slower.

For example, a 12.5 MB  file took less than a second to compress on the older version, but it took about 15 seconds on the latest beta version.

The newest version does write out a bunch of logging information that wasn't in the previous version:

```
[General] Calculated new CRC: 4258584823 from old CRC: 4193573647
[General] Writing 1024 bytes into archive from buffer
[General] Calculated new CRC: 1261850499 from old CRC: 4258584823
[General] Writing 1024 bytes into archive from buffer
[General] Calculated new CRC: 0500446323 from old CRC: 1261850499
[General] Writing 1024 bytes into archive from buffer
```

Could it be calculating the new CRC after every write is slowing it down?

**Comment by abbeycode** [comment]
It’s not doing anything with the CRC that it wasn’t before. What method are you calling, and is it possible to include the archive (or email it to me?)

**Comment by brendand** [comment]
I'm using the following bit of code to write out an archive:

```
+ (BOOL)zipItemAtURL:(NSURL *)itemURL archiveURL:(NSURL *)archiveURL progress:(NSProgress *)compressProgress completion:(nullable void (^)(NSProgress * __nullable progress))completionHandler error:(NSError **)error {
	
	NSError *localError = nil;
	
	BOOL success = NO;

	UZKArchive *archive = [[UZKArchive alloc] initWithURL:archiveURL
													error:&localError];
	
	NSFileManager *fileManager = [[NSFileManager alloc] init];
	
	NSArray *subPaths = [fileManager subpathsOfDirectoryAtPath:itemURL.path
													  error:&localError];
	
	NSProgress *progress = [[NSProgress alloc] init];
	progress.totalUnitCount = subPaths.count;
	progress.completedUnitCount = 0;
	
	for (NSString *subPath in subPaths) {
		
		if (compressProgress.isCancelled) {
			if (completionHandler) {
				completionHandler(progress);
			}
			return YES;
		}
		
		NSString *filePath = [[itemURL path] stringByAppendingPathComponent:subPath];
		
		NSURL *fileURL = [NSURL fileURLWithPath:filePath];
		
		NSString *filePathInZip = [[itemURL lastPathComponent] stringByAppendingPathComponent:subPath];
		
		NSNumber *isDirectory = nil;
		[fileURL getResourceValue:&isDirectory forKey:NSURLIsDirectoryKey error:NULL];
		
		NSDate *modificationDate = nil;
		[fileURL getResourceValue:&modificationDate forKey:NSURLAttributeModificationDateKey error:&localError];
		
		// skip directories since the directory will effectively be in the archive
		// when we add a real file to it with a subPath that contains the directory path.
		
		if (![isDirectory boolValue]) {
			
			NSData *fileData = [NSData dataWithContentsOfURL:fileURL
													 options:NSDataReadingUncached | NSDataReadingMappedIfSafe
													   error:&localError];
			const void *bytes = fileData.bytes;
			unsigned int bufferSize = 1024; //Arbitrary
			NSError *writeError = nil;
			
			success = [archive writeIntoBuffer:filePathInZip
					fileDate:modificationDate
					error:&writeError
					block:
			^BOOL(BOOL(^writeData)(const void *bytes, unsigned int length), NSError**(actionError)) {
			     for (NSUInteger i = 0; i <= fileData.length; i += bufferSize) {
			          unsigned int length = (unsigned int)MIN(fileData.length - i, bufferSize);
			          if (!writeData(&bytes[i], length)) {
			                 return NO;
			          }
			     }
						   
			 return YES;
			}];
			
			if (success) {
				progress.completedUnitCount += 1;
				progress.localizedAdditionalDescription = [filePath lastPathComponent];
				if (completionHandler) {
					completionHandler(progress);
				}
			} else {
				TFFLog(@"Error writing zip file: %@", writeError);
			}
		}
	}
	
	*error = localError;
	
	return success;
}
```

**Comment by brendand** [comment]
And for extracting it seems I'm using a deprecated API:

```
		success = [oldArchive extractFilesTo:NSTemporaryDirectory()
								   overwrite:YES
									progress:^(UZKFileInfo * _Nonnull currentFile, CGFloat percentArchiveDecompressed) {
										dispatch_async(dispatch_get_main_queue(), ^{
											float percent = percentArchiveDecompressed * 100;
											progress.completedUnitCount = percent;
											progress.localizedAdditionalDescription = [currentFile.filename lastPathComponent];
										});
									} error:&error];
```

I'll work on converting it to the recommended API.

**Comment by abbeycode** [comment]
I deprecated that method because you don't need a `progress` block anymore - I support `NSProgress` and `NSProgressReporting`, which is what it looks like you're already using. You can use it implicitly or explicitly, with a guide on the README for the `1.9` branch.

To help me reproduce your issue, since I don't see anything blatantly wrong with your code, can you provide an example file you're compressing or decompressing that's slower?

Out of curiosity, have you run in Instruments to profile where it's spending its time?

**Comment by brendand** [comment]
Sure, here's a file that I compressed:

https://www.dropbox.com/s/ddtgoajuj5iwg2k/Sample.tapforms-2017-11-09T19-12-36.zip?dl=0

It's 12.4 MB compressed.

It's 13.4 MB uncompressed.

With the older library it took 1.073878 seconds. With the new version it took 97.259650 seconds to compress the same document file. So almost a hundred times slower.

However, when I ran it in Instruments, it was about the same speed as the previous version. So maybe the issue is all the logging it's doing within Xcode. I tried an archive build of my app too and that also performed well.

So it really must just be the logging causing the performance problem from Xcode then. How do I turn off the logging? Is there an environment variable I can use?
